### PR TITLE
Fastrpc fixes

### PR DESCRIPTION
**Pull request by dsharletg**
This PR fixes a number of issues with FastRPC:

- The rpcmem code in the Hexagon SDK is really hard to understand and, IMO, very bug prone. To rule out a bug, I rewrote it. The new implementation is contained entirely in halide_host.cpp in the hexagon runtime, so we don't need to deal with it in the Halide runtime.

- While it wasn't causing any issues, it's technically not good to have a non-zero return values from a FastRPC call, so modify get_symbol to return the symbol in an output parameter instead. This PR adds a shim to the host side runtime for seamless compatibility.

### CODE REVIEW

**Comment by dsharletg** [comment]
Hmm, somehow the prebuilts got lost in the merge commit. I will re-add them.

File: src/runtime/hexagon_remote/halide_host.cpp [hunk 1]
```diff
@@ -1,23 +1,237 @@
...
+#include <sys/mman.h>
+#include <linux/types.h>
+#include <android/log.h>
+
+#include "bin/src/halide_hexagon_remote.h"
+
+typedef halide_hexagon_remote_handle_t handle_t;
+typedef halide_hexagon_remote_buffer buffer;
+
+typedef int ion_user_handle_t;
```
Comment (zvookin) [inline src/runtime/hexagon_remote/halide_host.cpp]: I assume these are mirroring data structures from Ion. Can we add a comment explaining this and perhaps what Ion is?

File: src/runtime/hexagon_remote/halide_host.cpp [hunk 2]
```diff
@@ -1,23 +1,237 @@
...
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <linux/types.h>
+#include <android/log.h>
+
+#include "bin/src/halide_hexagon_remote.h"
+
+typedef halide_hexagon_remote_handle_t handle_t;
```
Comment (zvookin) [inline src/runtime/hexagon_remote/halide_host.cpp]: It seems like there should be some namespacing on these two declarations.

File: src/runtime/hexagon_remote/halide_host.cpp [hunk 3]
```diff
@@ -1,23 +1,237 @@
...
+}
+
+int ion_free(int ion_fd, ion_user_handle_t ion_handle) {
+    if(ioctl(ion_fd, ION_IOC_FREE, &ion_handle) < 0) {
+        return -1;
+    }
+    return 0;
+}
+
+struct allocation_record {
```
Comment (zvookin) [inline src/runtime/hexagon_remote/halide_host.cpp]: Internal to implementation? Can go in a namespace?

File: src/runtime/hexagon_remote/halide_host.cpp [hunk 4]
```diff
@@ -1,23 +1,237 @@
...
+    int ion_fd = open("/dev/ion", O_RDONLY, 0);
+    if (ion_fd < 0) {
+        __android_log_print(ANDROID_LOG_ERROR, "halide", "open('/dev/ion') failed");
+        return NULL;
+    }
+
+    const int heap_id = 25;  // system heap
+    const int ion_flags = 1;  // cached
+
+    // Align the size up to pages.
```
Comment (zvookin) [inline src/runtime/hexagon_remote/halide_host.cpp]: "Align the size up to 4k pages."?

File: src/runtime/hexagon_remote/halide_host.cpp [hunk 5]
```diff
@@ -1,23 +1,242 @@
...
+
+namespace {
+
+typedef halide_hexagon_remote_handle_t handle_t;
+typedef halide_hexagon_remote_buffer buffer;
+
+typedef int ion_user_handle_t;
+
+struct ion_allocation_data {
+	size_t len;
```
Comment (steven-johnson) [inline src/runtime/hexagon_remote/halide_host.cpp]: did you mean to use tabs (vs. spaces) here and elsewhere?

File: src/runtime/hexagon_remote/halide_host.cpp [hunk 6]
```diff
@@ -1,23 +1,242 @@
...
+    allocation_record *next;
+    int ion_fd;
+    ion_user_handle_t handle;
+    int buf_fd;
+    void *buf;
+    size_t size;
+};
+
+// Make a dummy allocation so we don't need a special case for the head node.
+allocation_record allocations = { NULL, };
```
Comment (steven-johnson) [inline src/runtime/hexagon_remote/halide_host.cpp]: Just using = {} should zero all fields.

File: src/runtime/hexagon_remote/halide_host.cpp [hunk 7]
```diff
@@ -1,23 +1,253 @@
...
-    rpcmem_free(ptr);
+    if (!ptr) {
+        return;
+    }
+
+    // Find the record for this allocation and remove it from the list.
+    pthread_mutex_lock(&allocations_mutex);
+    allocation_record *rec = &allocations;
+    while (rec) {
+        if (rec && rec->next->buf == ptr) {
```
Comment (dsharletg) [inline src/runtime/hexagon_remote/halide_host.cpp]: Just spotted a bug here, will fix

**Comment by dsharletg** [comment]
This *might* be unstable (though it doesn't appear to be a memory leak). Going to close this for now since the bug it attempted to work around was resolved anyways.

**Pull request closed without merging**